staging/src/k8s.io/apiserver/pkg/header/context.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package header
+
+import (
+	"context"
+	"net/http"
+	"sync"
+)
+
+// The key type is unexported to prevent collisions
+type key int
+
+const (
+	// headerAccessorKey is the context key for the header accessor.
+	headerAccessorKey key = iota
+)
+
+type Header interface {
+	// Add calls header.Add(key,value) if not frozen
+	Add(key, value string)
+	// Set calls header.Set(key,value) if not frozen
+	Set(key, value string)
+	// Del calls header.Del(key) if not frozen
+	Del(key string)
+	// Freeze prevents future Add/Set/Del calls from modifying headers
+	Freeze()
+}
+
+type header struct {
+	mu     sync.Mutex
+	frozen bool
+	header http.Header
+}
+
+func (h *header) Add(key, value string) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if !h.frozen {
```

Comment:
Do we need to warn (higher debug levels?) when we drop these?

Reply:
I don't think so? the only time that happens is in a timeout and I don't think log messages about headers being dropped are useful at all

I actually tweaked this from a "freeze" pattern to a "detach" pattern so calls to add/set/del still *do* things but no longer propagate to the shared header set. That lets things holding the safe header setter remain internally consistent even if they aren't propagating